router/proxy/transport.go

```diff
@@ -64,13 +66,33 @@ func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend s
 	}
 }
 
-// always sets the response.Request.URL.Host to the last backend that was
-// connected to.
-func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
 	// http.Transport closes the request body on a failed dial, issue #875
 	req.Body = &fakeCloseReadCloser{req.Body}
 	defer req.Body.(*fakeCloseReadCloser).RealClose()
 
+	resultc := make(chan httpResult, 1)
+
+	go func() {
+		defer close(resultc)
+		res, err := t.roundTrip(req)
+		resultc <- httpResult{res, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		httpTransport.CancelRequest(req)
+		err := ctx.Err()
+		if err == context.Canceled {
+			return nil, nil
```

Comment:
IMO it would be safer to return an `errRequestCanceled` here. It's correct right now, but in the future if the context could be canceled by something other than the `ReverseProxy`, then the `defer result.res.Body.Close()` in `ReverseProxy.ServeHTTP` would panic b/c `result.res` is nil.